auth/http.go

```diff
@@ -624,24 +624,85 @@ func httpHandlerLogout(central *ImqsCentral, w http.ResponseWriter, r *httpReque
 	identity, password, basicOK := r.http.BasicAuth()
 	if !basicOK {
 		authaus.HttpSendTxt(w, http.StatusBadRequest, authaus.ErrHttpBasicAuth.Error())
 		return
 	}
+
 	if identity == "" {
 		httpSendNoIdentity(w)
 		return
 	}
 
+	// Attempt login, returning session if valid
 	sessionKey, token, err := central.Central.Login(identity, password, getIPAddress(r.http))
 	if err != nil {
-		authaus.HttpSendTxt(w, http.StatusForbidden, err.Error())
-		if user, eUser := central.Central.GetUserFromIdentity(identity); eUser == nil {
+		// This could still be a passthrough login to MSAAD - retain username/password
+		// Requestor should in this case provide client_id for us to validate
+
+		user, eUser := central.Central.GetUserFromIdentity(identity)
+		if eUser != nil {
+			authaus.HttpSendTxt(w, http.StatusUnauthorized, err.Error())
+			return
+		}
+
+		// user exists, continue
+		r.http.ParseForm()
+		clientId := r.http.Form.Get("client_id")
+		tryPassThrough := clientId != ""
+
+		// if client_id is blank or not present, assume normal login and FAIL
+		if !tryPassThrough {
+			authaus.HttpSendTxt(w, http.StatusUnauthorized, err.Error())
 			auditUserLogAction(central, r, user.UserId, identity, "User Profile: "+identity, authaus.AuditActionFailedLogin)
+			return
 		}
-		return
+
+		tryPassThrough = false
+		// validate against whitelist
+		for _, cid := range central.Config.Authaus.MSAAD.PassthroughClientIDs {
+			if clientId == cid {
+				tryPassThrough = true
+				break
+			}
+		}
+		if !tryPassThrough {
+			whiteListFailedMsg := fmt.Sprintf("Passthrough auth: client_id specified [%v] is not whitelisted. \n", clientId)
+			central.Central.Log.Warn(whiteListFailedMsg)
+			authaus.HttpSendTxt(w, http.StatusForbidden, whiteListFailedMsg)
+			return
+		}
+
+		if user.Type != authaus.UserTypeMSAAD {
+			notMSAADUserMsg := "Not an MSAAD user."
+			central.Central.Log.Warn(notMSAADUserMsg)
+			authaus.HttpSendTxt(w, http.StatusUnauthorized, notMSAADUserMsg)
+			return
+		}
+
+		// try MSAAD login
+		if e2 := central.Central.OAuth.OAuthLoginUsernamePassword(r.http, identity, password); e2 != nil {
+			central.Central.Log.Warn("MSAAD login failed")
+			authaus.HttpSendTxt(w, http.StatusUnauthorized, e2.Error())
+			return
+		}
+
+		// At this point the user may not have an IMQS Auth session, since MSAAD sync only populates the authusergroup table.
+		// It will, however, eventually have a permit, which would have come from the MSAAD background sync if permissions have been assigned.
+		// So we try to create a session.
+		// Authaus requires an oauthSession as a primary key, but we don't get one back from using username password
+		// authentication, so we just use the user email for now
+		sessionKey2, token2, err := central.Central.CreateSession(&user, r.http.RemoteAddr, user.Email)
+		if err != nil {
+			authaus.HttpSendTxt(w, http.StatusForbidden, err.Error())
+			return
+		}
+		sessionKey = sessionKey2
+		token = token2
```

Comment:
I think everything from this point should be moved into a function (not sure which file... probably this one, even thou this file is way to big).

It should take `r *httpRequest`, `user *AuthUser` and `password string` as parameters. No need for identity, you can pull the username from the AuthUser struct (or email, if we really really need it).

It should return `sessionKey string`, `token *authhaus.Token`, `httpCode int` and `err Error`. And just have them replace the current variables. You can leave your logging in the function, but the `authaus.HttpSendTxt(...` should definitely still happen here.

Note*: The only annoying thing about this for me is that we have to call `r.http.ParseForm()` in the function. Even thou we have already called it before the function, but we can't assume that who ever calls this function has already called it.